### ISSUE TITLE
zero is broken

### ISSUE THREAD

**Issue by thofma** [Nemocas/AbstractAlgebra.jl#407]
@rfourquet

The `zero` method
https://github.com/Nemocas/AbstractAlgebra.jl/blob/19b9b3859342518a14ee6effc982ca021621791f/src/generic/Matrix.jl#L220
introduced in #388 breaks everything in Nemo, since `zero` now returns objects of the wrong type:
```julia
julia> using Nemo;

julia> R = GF(5)
Galois field with characteristic 5

julia> z = matrix(R, 2, 2, [1, 2, 2, 3])
[1  2]
[2  3]

julia> typeof(z)
gfp_mat

julia> zero(z)
[0  0]
[0  0]

julia> typeof(ans)
AbstractAlgebra.Generic.MatSpaceElem{Nemo.gfp_elem}

julia> similar(z)
[0  0]
[0  0]

julia> typeof(ans)
gfp_mat

julia> similar(z, base_ring(z))
[#undef  #undef]
[#undef  #undef]

julia> typeof(ans)
AbstractAlgebra.Generic.MatSpaceElem{Nemo.gfp_elem}
```
I don't quite understand, why the `zero` methods have an additional base ring argument. Can we just delete this? I would like to tag new releases for all our packages to make life easier for @wbhart.

**Comment by rfourquet** [comment]
> introduced in #388 breaks everything in Nemo, since zero now returns objects of the wrong type:

Did it break something? If I'm not mistaken, `zero` was not previously defined on Nemo matrices, i.e. `zero(z)` from your example was throwing an error. That said, `zero` should indeed now be defined for Nemo matrices to return matrices in line with what `similar` does. The documentation states that "generally" it's not necessary to define `zero` for custom matrices, but I knew Nemo was an exception and was planning to fix that.

The `zero` method accepts an additional base ring argument to have the same signature as `similar`. I think it's not used, but I thought it could help the PR in which it was introduced to be accepted, so that `zero` can be a drop-in replacement to `similar`. As the base ring argument was added to `similar` very recently, I believe no-one uses it yet (I use it in a PR in preparation though).

**Comment by thofma** [comment]
```julia
julia> R, = FiniteField(fmpz(5), 1, "a")
(Finite field of degree 1 over F_5, 0)

julia> Rt, t = R["t"]
(Univariate Polynomial Ring in t over Finite field of degree 1 over F_5, t)

julia> z = matrix(R, 2, 2, [1, 2, 2, 3])
[1  2]
[2  3]

julia> subst(charpoly(Rt, z), z)
ERROR: MethodError: Cannot `convert` an object of type AbstractAlgebra.Generic.MatSpaceElem{fq} to an object of type fq_mat
Closest candidates are:
  convert(::Type{T}, ::T) where T at essentials.jl:167
Stacktrace:
 [1] setindex!(::Array{fq_mat,1}, ::AbstractAlgebra.Generic.MatSpaceElem{fq}, ::Int64) at ./array.jl:766
 [2] powers(::fq_mat, ::Int64) at /home/thofmann/.julia/dev/AbstractAlgebra/src/generic/Matrix.jl:758
 [3] subst(::fq_poly, ::fq_mat) at /home/thofmann/.julia/dev/AbstractAlgebra/src/generic/Poly.jl:2617
 [4] top-level scope at REPL[46]:1
```
It is not clear what `zero` should do with an additional base ring argument. Isn't it then just `zero_matrix`?

**Comment by wbhart** [comment]
The extra ring argument is probably not needed, as the zero function can just get the base ring. But presumably @rfourquet prefers to make it a default argument so that he can reuse the same function with different rings for some other purpose (the change_base_ring stuff, for example)?

**Comment by rfourquet** [comment]
Ok nice example, actually #388 broke `eye` for Nemo matrices, as `eye` uses now `zero` instead of `similar`.

`zero(mat, ring, ...)` with an additional base ring is not exactly equivalent to `zero_matrix`, it's equivalent to `zero!(similar(mat, ring, ...)`, so really it's a convenience method. But I admit I don't have a use-case yet, unlike for `similar` with a specified ring.

**Comment by thofma** [comment]
Isn't `zero!(similar(mat, ring, ...)` doing the same as `zero_matrix`? I mean functionality wise, not the same code path.

**Comment by rfourquet** [comment]
> Isn't zero!(similar(mat, ring, ...) doing the same as zero_matrix? I mean functionality wise, not the same code path.

It's the same difference you explained to me yourself [there](https://github.com/Nemocas/AbstractAlgebra.jl/pull/388#issuecomment-532556973) :)
```julia
julia> A = one(Generic.MatSpace{fmpz}(FlintZZ, 2, 2, false))
[1  0]
[0  1]

julia> typeof(A)
AbstractAlgebra.Generic.MatSpaceElem{fmpz}

julia> typeof(zero_matrix(base_ring(A), 2, 3))
fmpz_mat

julia> typeof(zero(A, base_ring(A), 2, 3)) # == typeof(zero!(similar(A, base_ring(A), 2, 3)))
AbstractAlgebra.Generic.MatSpaceElem{fmpz}
```

**Comment by thofma** [comment]
OK, right :)

OK, let me rephrase. Here is an example which shows that the current behavior is non-optimal.
```julia
julia> z = matrix(FlintZZ, 2, 2, [1, 2, 2, 3])
[1  2]
[2  3]

julia> typeof(ans)
fmpz_mat

julia> zero_matrix(base_ring(z), 2, 2)
[0  0]
[0  0]

julia> similar(z, base_ring(z), 2, 2)
[#undef  #undef]
[#undef  #undef]

julia> typeof(ans)
AbstractAlgebra.Generic.MatSpaceElem{fmpz}

julia> R = GF(5)
Galois field with characteristic 5

julia> similar(z, R, 2, 2)
[#undef  #undef]
[#undef  #undef]

julia> typeof(ans)
AbstractAlgebra.Generic.MatSpaceElem{Nemo.gfp_elem}

julia> zero_matrix(R, 2, 2)
[0  0]
[0  0]

julia> typeof(ans)
gfp_mat
```
So at the moment whenever `similar` or `zero` have a ring as a second argument, it will create a generic matrix. I don't think that we want this. In particular that `similar(z, base_ring(z), nrows(z), ncols(z)) != similar(z, nrows(z), ncols(z))`.

I actually don't think that `similar` and `zero` should have a second ring argument at all. What they produce is independent of the type of the input matrix (the first argument). Moreover, what they are supposed to do is already there: If one wants a matrix over a new ring with given dimensions, one should just use `similar(ring,..)` and `zero(ring,...)`.

So I am proposing to remove the second ring argument. It produces inconsistent results and adds no functionality. It is actually very confusing.

**Comment by thofma** [comment]
No, by ring I mean `base_ring(z)`, not the parent of `z` (I think you use the same convention).

No, generic matrices don't play the role of `AbstractArray`. If I am using Nemo or any other package which implements a matrix type, then this is not true.

I don't understand the last point. `similar(::Generic.MatElem{fmpz})` was already returning a `Generic.MatElem{fmpz}` and `zero_matrix` was already giving back an `fmpz_mat`.

I understand what the code is supposed to do. I am just saying that in the end, if you want to fix the issues raised above, `similar(mat, ring,...)` will be the same as `zero_matrix(base_ring(r), nrows(r), ncols(c))` (with the possibility of uninitialized entries). The reason is simply that the return type of `similar(mat, ring,..)` must be the return type of `matrix(ring,...)` and the only functions that can be used for this within AA are `matrix` or `zero_matrix`.

**Comment by wbhart** [comment]
I don't know of any reason to not make a call out to parent. Perhaps you are thinking of our desire to avoid *constructing* parent objects in library code (and when such constructions are made it should be via a call to AbstractAlgebra.NameOfParentConstructor rather than just to NameOfParentConstructor).

I guess calling parent objects may not be efficient in Julia, but I am not sure, and we use it everywhere, I think.

**Comment by thofma** [comment]
The only reason `similar` exists is to remove all calls to parent for matrices.

**Comment by rfourquet** [comment]
> one should just use similar(ring,..) and zero(ring,...).

> by ring I mean base_ring(z)

Ok, I was confused because these functionalities (`similar(ring, ...)` and `zero(ring, ...)`) to create a matrix don't exist currently. On the other hand, `zero(parent(z))` does exist and creates a `zero` matrix. So `zero(R::Ring)` creates a zero object belonging to the passed ring `R`, so it would be too confusing to have `zero(R, 2, 3)` create a zero matrix with base ring `R`.

> No, generic matrices don't play the role of AbstractArray. 

I meant `Array`, but from a certain point of vue, it does in my opinion: `Array` can hold elements of any type, like generic AA matrices can hold element from any ring, whereas `Nemo` matrices can hold elements from a limited number of rings. So AA matrices will always work as a fall-back.

> If I am using Nemo or any other package which implements a matrix type, then this is not true.

I don't really understand. Nemo or another package depending on AA will always have access to AA matrices, like any package in Julia can use `Array`.

> I don't understand the last point. similar(::Generic.MatElem{fmpz}) was already returning a Generic.MatElem{fmpz} and zero_matrix was already giving back an fmpz_mat.

I was trying to understand which already existing functionality makes `similar(mat, ring, ...)`  redundant, the closest I find is `zero_matrix`. 

> if you want to fix the issues raised above, `similar(mat, ring,...)` will be the same as `zero_matrix(ring, ...)` 

No, if `mat` is a generic AA matrix and `ring` a Nemo ring (say `fmpz`), then `similar(mat, ring)` will give a generic AA matrix, while `zero_matrix` will give `fmpz_mat`.

> The reason is simply that the return type of similar(mat, ring,..) must be the return type of matrix(ring,...)

I don't understand... `matrix(FlintZZ, ...)` creates a `fmpz_mat` whereas `similar(mat, FlintZZ, ...)` can create a generic AA matrix provided `mat` is.

I should soon finish my fix, which is against Nemo and not AbstractAlgebra, and we will have something more concrete to discuss.

**Comment by thofma** [comment]
> Ok, I was confused because these functionalities (`similar(ring, ...)` and `zero(ring, ...)`) to create a matrix don't exist currently. On the other hand, `zero(parent(z))` does exist and creates a `zero` matrix. So `zero(R::Ring)` creates a zero object belonging to the passed ring `R`, so it would be too confusing to have `zero(R, 2, 3)` create a zero matrix with base ring `R`.

My bad, I meant `zero_matrix`, sorry.

> > if you want to fix the issues raised above, `similar(mat, ring,...)` will be the same as `zero_matrix(ring, ...)`
> 
> No, if `mat` is a generic AA matrix and `ring` a Nemo ring (say `fmpz`), then `similar(mat, ring)` will give a generic AA matrix, while `zero_matrix` will give `fmpz_mat`.

Which is exactly what we want, see below.

> > The reason is simply that the return type of similar(mat, ring,..) must be the return type of matrix(ring,...)
> 
> I don't understand... `matrix(FlintZZ, ...)` creates a `fmpz_mat` whereas `similar(mat, FlintZZ, ...)` can create a generic AA matrix provided `mat` is.
> 
> I should soon finish my fix, which is against Nemo and not AbstractAlgebra, and we will have something more concrete to discuss.

Something that we definitely don't want is that `similar(::fmpz_mat, ::FlintRationalField)` creates a generic matrix. It should create whatever `zero_matrix(::FlintRationalField,..)` or `matrix(FlintRationalField,...)` would create.

If I do `using Nemo; z = zero_matrix(ZZ, 2, 2)` then `similar(z, FlintQQ)` should create an `fmpq_mat` and not a generic matrix. If I have `z` a `Generic.MatElem{fmpz}`, then I want `similar(z, FlintQQ)` to be a `Generic.MatElem{fmpq}`.

I have a feeling you are trying to "fix" something, which we consider as the behavior that we want.

**Comment by rfourquet** [comment]
> If I do `using Nemo; z = zero_matrix(ZZ, 2, 2)` then `similar(z, FlintQQ)` should create an `fmpq_mat` and not a generic matrix. If I have `z` a `Generic.MatElem{fmpz}`, then I want `similar(z, FlintQQ)` to be a `Generic.MatElem{fmpq}`.

Ok so we seem to agree :) 
This is exactly what my fix is about. 

> I have a feeling you are trying to "fix" something, which we consider as the behavior that we want.

My "fix" is exactly to implement the behavior you laid out above, i.e. currently `similar(z::fmpz_mat, FlintQQ)` returns a generic matrix, whereas we want it to return an `fmpq_mat`.
